virtcontainers/clh.go

```diff
@@ -1202,141 +1175,195 @@ func (o *CLICmdline) Build(cmdline *CommandLine) {
+	if err != nil {
+		return openAPIClientError(err)
 	}
 
-	if d.apiSocket != nil {
-		d.apiSocket.Build(cmdLine)
-	}
+	info, _, err := cl.VmInfoGet(ctx)
 
-	if d.logFile != nil {
-		d.logFile.Build(cmdLine)
+	if err != nil {
+		return openAPIClientError(err)
 	}
 
-	if d.cpus != nil {
-		d.cpus.Build(cmdLine)
-	}
-	if d.memory != nil {
-		d.memory.Build(cmdLine)
-	}
-	if d.disk != nil {
-		d.disk.Build(cmdLine)
-	}
-	if d.rng != nil {
-		d.rng.Build(cmdLine)
+	clh.Logger().Debugf("VM state after create: %#v", info)
+
+	if info.State != "Created" {
+		return fmt.Errorf("VM state is not 'Created' after 'CreateVM'")
 	}
-	if d.vsock != nil {
-		d.vsock.Build(cmdLine)
+
+	clh.Logger().Debug("Booting VM")
+	_, err = cl.BootVM(ctx)
+
+	if err != nil {
+		return openAPIClientError(err)
 	}
-	if d.fs != nil {
-		d.fs.Build(cmdLine)
+
+	info, _, err = cl.VmInfoGet(ctx)
+
+	if err != nil {
+		return openAPIClientError(err)
 	}
-	if d.kernel != nil {
-		d.kernel.Build(cmdLine)
+
+	clh.Logger().Debugf("VM state after boot: %#v", info)
+
+	if info.State != "Running" {
+		return fmt.Errorf("VM state is not 'Running' after 'BootVM'")
 	}
-	if d.nets != nil {
-		d.nets.Build(cmdLine)
+
+	return nil
+}
+
+func (clh *cloudHypervisor) addVSock(cid int64, path string) {
+	clh.Logger().WithFields(log.Fields{
+		"path": path,
+		"cid":  cid,
+	}).Info("Adding HybridVSock")
+
+	clh.vmconfig.Vsock = append(clh.vmconfig.Vsock, chclient.VsockConfig{Cid: cid, Sock: path})
+}
+
+func (clh *cloudHypervisor) addNet(e Endpoint) {
+	clh.Logger().WithField("endpoint-type", e).Debugf("Adding Endpoint of type %v", e)
+	mac := e.HardwareAddr()
+	tapPath := e.NetworkPair().TapInterface.TAPIface.Name
+	clh.Logger().WithFields(log.Fields{
+		"mac": mac,
+		"tap": tapPath,
+	}).Info("Adding Net")
+
+	// FIXME: This is required by CH
+	// remove after PR is merged:
+	// https://github.com/cloud-hypervisor/cloud-hypervisor/pull/480
+	ip := "0.0.0.0"
+	mask := "0.0.0.0"
+	clh.vmconfig.Net = append(clh.vmconfig.Net, chclient.NetConfig{Mac: mac, Tap: tapPath, Ip: ip, Mask: mask})
+}
+
+// Add shared Volume using virtiofs
+func (clh *cloudHypervisor) addVolume(volume types.Volume) error {
+	if clh.config.SharedFS != config.VirtioFS {
+		return fmt.Errorf("shared fs method not supported %s", clh.config.SharedFS)
 	}
-	if d.cmdline != nil {
-		d.cmdline.Build(cmdLine)
+
+	vfsdSockPath, err := clh.virtioFsSocketPath(clh.id)
+	if err != nil {
+		return err
 	}
 
-	return cmdLine, nil
-}
+	if clh.config.VirtioFSCache == virtioFsCacheAlways {
+		clh.vmconfig.Fs = []chclient.FsConfig{
+			{
+				Tag:       volume.MountTag,
+				CacheSize: int64(clh.config.VirtioFSCacheSize << 20),
+				Sock:      vfsdSockPath,
+				NumQueues: clhFsQueues,
+				QueueSize: clhFsQueueSize,
+			},
+		}
+	} else {
+		clh.vmconfig.Fs = []chclient.FsConfig{
+			{
+				Tag:       volume.MountTag,
+				Sock:      vfsdSockPath,
+				NumQueues: clhFsQueues,
+				QueueSize: clhFsQueueSize,
+			},
+		}
```

Comment:
I'm not sure about the behavior from this part. By not providing the `CacheSize`, and because the OpenAPI says that it expects an `int64`, I think the default value is going to be 0. But that will be translated into `Some(0)` for the Rust code, meaning it will still try to use DAX with the shared memory region when really what you want is no cache at all.
Please give it a try and let me know if that's the case, but TBH I think that's something we need to fix in CH.

Reply:
@jcvenegas I've just opened https://github.com/cloud-hypervisor/cloud-hypervisor/pull/515 to fix this. It'd be nice if you could give it a shot as I think this is the right way to configure `FsConfig`.
Note that because I added `dax` default to `true` and `cache_size` default to `8G`, you shouldn't need to specify `dax` and `cache_size` when the only information you get from your user is `virtioFsCacheAlways   = "always"`.